actors/builtin/miner/miner_test.go

```diff
@@ -899,53 +1057,509 @@ func TestCommitments(t *testing.T) {
+			expectedPowerDelta: expectedPower,
+			expectedPenalty:    big.Zero(),
+		})
+
+		// replaced sector expires at cron, removing its power and pledge.
+		expectedPowerDelta := oldPower.Neg()
+		actor.onDeadlineCron(rt, &cronConfig{
+			expiredSectorsPowerDelta:  &expectedPowerDelta,
+			expiredSectorsPledgeDelta: oldSector.InitialPledge.Neg(),
+			expectedEnrollment:        rt.Epoch() + miner.WPoStChallengeWindow,
+		})
+	})
+
+	t.Run("declare fault for replaced cc upgrade sector doesn't double subtract power", func(t *testing.T) {
+		actor := newHarness(t, periodOffset)
+		rt := builderForHarness(actor).
+			WithBalance(bigBalance, big.Zero()).
+			Build(t)
+		actor.constructAndVerify(rt)
+
+		oldSector, newSector := actor.commitProveAndUpgradeSector(rt, 100, 200, defaultSectorExpiration, []abi.DealID{1})
+
+		st := getState(rt)
+		dlIdx, partIdx, err := st.FindSector(rt.AdtStore(), oldSector.SectorNumber)
+		require.NoError(t, err)
+
+		// Both sectors are present (in the same deadline/partition).
+		deadline, partition := actor.getDeadlineAndPartition(rt, dlIdx, partIdx)
+		assert.Equal(t, uint64(2), deadline.TotalSectors)
+		assert.Equal(t, uint64(2), deadline.LiveSectors)
+
+		// The old sector's expiration has changed to the end of this proving deadline.
+		// The new one expires when the old one used to.
+		// The partition is registered with an expiry at both epochs.
+		pIdx := uint64(0)
+		dQueue := actor.collectDeadlineExpirations(rt, deadline)
+		dlInfo := miner.NewDeadlineInfo(st.ProvingPeriodStart, dlIdx, rt.Epoch())
+		quantizedExpiration := miner.QuantSpecForDeadline(dlInfo).QuantizeUp(oldSector.Expiration)
+		assert.Equal(t, map[abi.ChainEpoch][]uint64{
+			dlInfo.NextNotElapsed().Last(): {pIdx},
+			quantizedExpiration:            {pIdx},
+		}, dQueue)
+
+		pQueue := actor.collectPartitionExpirations(rt, partition)
+		assertBitfieldEquals(t, pQueue[dlInfo.NextNotElapsed().Last()].OnTimeSectors, uint64(oldSector.SectorNumber))
+		assertBitfieldEquals(t, pQueue[quantizedExpiration].OnTimeSectors, uint64(newSector.SectorNumber))
+
+		// now declare old sector faulty
+		actor.declareFaults(rt, oldSector)
+
+		deadline, _ = actor.getDeadlineAndPartition(rt, dlIdx, partIdx)
+		dQueue = actor.collectDeadlineExpirations(rt, deadline)
+		dlInfo = miner.NewDeadlineInfo(st.ProvingPeriodStart, dlIdx, rt.Epoch())
+		expectedReplacedExpiration := miner.QuantSpecForDeadline(dlInfo).QuantizeUp(rt.Epoch() + miner.FaultMaxAge)
+		quantizedExpiration = miner.QuantSpecForDeadline(dlInfo).QuantizeUp(oldSector.Expiration)
+
+		// deadling marks expirations for partition at expiration epoch
+		assert.Equal(t, map[abi.ChainEpoch][]uint64{
+			dlInfo.NextNotElapsed().Last(): {pIdx},
+			expectedReplacedExpiration:     {pIdx},
+			quantizedExpiration:            {pIdx},
+		}, dQueue)
+
+		// but partitions expiration set at that epoch is empty
```

Comment:
Why is it empty?

Reply:
Shouldn't `DeclareFaults` place the old sector in `EarlySectors` at the fault expiration epoch?

Reply:
Since the sector had an on time expiration earlier than the fault expiration, it is left in its original expiration set, but its power is moved from active to faulty. This is checking that the expiration set at the fault expiration is empty, indicating it was not moved. Prior to this PR, it would have been moved to this expiration set and left in its original expiration set causing it to expire twice.